### PR TITLE
Make WalletInitInterface and DummyWalletInit private, fix nullptr deref

### PR DESCRIPTION
**Pull request by promag**
Implementations of `WalletInitInterface` don't have to be public, so make them private. This makes the interface instantiation static. Also reduces `ENABLE_WALLET` usage and removes the unnecessary `src/wallet/init.h` header.

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -72,7 +72,24 @@ static const bool DEFAULT_STOPAFTERBLOCKIMPORT = false;
...
+    void RegisterRPC(CRPCTable &) override {}
+    bool Verify() override {return true;}
+    bool Open() override {return true;}
+    void Start(CScheduler& scheduler) override {}
+    void Flush() override {}
+    void Stop() override {}
+    void Close() override {}
+};
+
+std::unique_ptr<WalletInitInterface> g_wallet_init_interface(new DummyWalletInit);
```
Comment (ajtowns) [inline src/init.cpp]: Could make this `WalletInitInterface* const g_wallet_init_interface(new DummyWalletInit);`

**Comment by jnewbery** [comment]
This change fixes the nullpointer dereference here: https://github.com/bitcoin/bitcoin/commit/c7ec5243892c38f9f77781b0e24a237942e7c776

**Comment by jamesob** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/12836/commits/208fad6b5e6c5f19a917e0dd06a75927281f31eb

File: src/init.cpp [hunk 2]
```diff
@@ -72,7 +72,25 @@ static const bool DEFAULT_STOPAFTERBLOCKIMPORT = false;
...
+    void RegisterRPC(CRPCTable &) override {}
+    bool Verify() override {return true;}
+    bool Open() override {return true;}
+    void Start(CScheduler& scheduler) override {}
+    void Flush() override {}
+    void Stop() override {}
+    void Close() override {}
+};
+
+static DummyWalletInit g_dymmy_wallet_init;
```
Comment (jnewbery) [inline src/init.cpp]: typo `dymmy`.

Why does this need to be static?

Comment (promag) [reply src/init.cpp]: Dummy typo 😄It doesn't have to be static. IMO these can have internal linkage. WDYT?

Comment (jnewbery) [reply src/init.cpp]: >  IMO these can have internal linkage. WDYT?

Yes, seems reasonable.

**Comment by jnewbery** [comment]
Tested ACK 57c00059bb51be64a670edebc00416fc2050d5f1 once this has been rebased on master to fix the linter error.

**Comment by promag** [comment]
Rebased.

**Comment by jamesob** [comment]
re-utACK https://github.com/bitcoin/bitcoin/pull/12836/commits/465a8eed157a637391f60d4681a5d7c6a885cdbd

**Comment by jnewbery** [comment]
Looks good. ReACK 465a8eed157a637391f60d4681a5d7c6a885cdbd

**Comment by instagibbs** [comment]
<MarcoFalke> ^ Needs rebase

File: src/init.cpp [hunk 3]
```diff
@@ -72,7 +72,25 @@ static const bool DEFAULT_STOPAFTERBLOCKIMPORT = false;
 
 std::unique_ptr<CConnman> g_connman;
 std::unique_ptr<PeerLogicValidation> peerLogic;
-std::unique_ptr<WalletInitInterface> g_wallet_init_interface;
+
+#if !(ENABLE_WALLET)
+class DummyWalletInit : public WalletInitInterface {
```
[the remark this answers is not in the document]
Comment (jamesob) [reply src/init.cpp]: I think the idea is to narrow the scope of this class to where it's needed and reduce unnecessary header bloat.

[the remark this answers is not in the document]
Comment (promag) [reply src/init.cpp]: Right.

**Comment by promag** [comment]
Rebased.

**Comment by jamesob** [comment]
Tested ACK https://github.com/bitcoin/bitcoin/pull/12836/commits/d894894aab78dbd5e32267313237648933248377

For both `--enable-wallet` and `--disable-wallet`:
  - [x] Ran `bitcoin-qt -h`
  - [x] Verified preexisting wallet data loaded

**Pull request merged**